### DISCUSSION TITLE
Separate handler state types

### DISCUSSION THREAD

**Discussion by danielgrigg** [tokio-rs/axum#1739] [Q&A]
Hello, new axum user and loving it!  I had a question on how to approach architecting routes and stateful dependencies. 

I'm trying to create some routes such that each handler injects only its required dependencies, eg a create_user handler might only inject a UserRepository but the whole app may consist of a hundred such repositories and bundling them all into a shared state isn't ideal.  
I saw there's a with_state method and the [docs and example](https://docs.rs/axum/latest/axum/routing/struct.Router.html#method.merge) suggest that independent handlers with differing state types can be merged together but I don't see it'd compile given the type constraint that the (using example types) InnerState be convertible into the OuterState? Am I trying to use it correctly? Or should I use another method?
For now I'm using a single state type but I can quickly see it becoming a god class coupled to everything.

**Comment by davidpdrsn** [comment]
> but I don't see it'd compile given the type constraint that the (using example types) InnerState be convertible into the OuterState? Am I trying to use it correctly? Or should I use another method?

You're probably missing `#[derive(FromRef)]` on your overall state type.

If you really do want different states you can do [this](https://docs.rs/axum/latest/axum/routing/struct.Router.html#merging-routers-with-state).
